### DISCUSSION TITLE
Generating ActiveX modules

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#85] [General]
_I guess it could be automated_:
The ActiveX API is stable except for the GUIDs, and LCIDs
I have hand modified the files to eliminate the need for users to create safe arrays by injecting converters into the function bodies 
So, when there’s an update from say AxApp24.py to AxApp25.py. I first auto generate the outline with the correct GUIDs, and LCIDs
Then winMerge in my changes, carefully lol. 
In short if they are reformatted it would be a disaster, I would need to revert changes to these files

_Originally posted by @CEXT-Dan in https://github.com/CEXT-Dan/PyRx/issues/78#issuecomment-2177349442_

**Comment by CEXT-Dan** [comment]
I think this area should be for bugs, so I can track them and closed them

**Comment by gswifort** [comment]
Check if it works.
If so, I'll try to optimize it.
[main.zip](https://github.com/user-attachments/files/15905778/main.zip)

**Reply by gswifort** [reply]
If at first glance it's not clear to you how to use it, don't bother - let me know and I'll explain it in the morning.  
In short, in the first run, import the module you improved, and in the second run, import the module from the new release that needs to be improved.

**Reply by CEXT-Dan** [reply]
Cool! I got an exception. But we can try this the other way around, these are the raw output files
AcadDBX24 and AxApp24, 

- extract only the DBX section of AcadDBX24 and insert it into AxApp24.
- Embed VT Arrays inside the function bodies 
- For each property in _prop_map_put_ that requires a VT array, Create an override 
- Fix the iterators that are not auto-generated correctly 
- Run the unit tests 

I also modify some of the Get functions so the user is not required to have "None" in the arguments 
I think there's only one or two of those

[AcadDBX24.zip](https://github.com/user-attachments/files/15906475/AcadDBX24.zip)

**Reply by gswifort** [reply]
How can I run the tests?

**Reply by CEXT-Dan** [reply]
You must have AutoCAD 2021-2024 installed, UnitTestPyActiveX.py is the test, its run from UnitTestRunner.py 
You actually don’t need to spend time doing this. These GUIDS only change every year, every three years on AutoCAD

I have a system to do this, copy the previous year and merge in the changes. 

I could have automated this whole project using SWIG, it’s not about that, this project is for me to keep busy in my retirement years. Otherwise I'd just have to go watch Netflix
